chainercv/links/model/vgg/vgg16.py

```diff
@@ -0,0 +1,157 @@
+import chainer
+from chainer.functions import dropout
+from chainer.functions import max_pooling_2d
+from chainer.functions import relu
+from chainer.functions import softmax
+from chainer.initializers import constant
+from chainer.initializers import normal
+
+from chainer.links import Linear
+
+from chainercv.utils import download_model
+
+from chainercv.links.connection.convolution_2d_block import Convolution2DBlock
+from chainercv.links.model.sequential_feature_extractor import \
+    SequentialFeatureExtractor
+
+
+# RGB order
+_imagenet_mean = np.array(
+    [123.68, 116.779, 103.939], dtype=np.float32)[:, np.newaxis, np.newaxis]
+
+
+class VGG16(SequentialFeatureExtractor):
+
+    """VGG16 Network for classification and feature extraction.
+
+    This model is a feature extraction link.
+    The network can choose to output features from set of all
+    intermediate and final features produced by the original architecture.
+    The value of :obj:`VGG16.feature_names` selects the features to be
+    collected by :meth:`__call__`.
+    :obj:`self.all_feature_names` is the list of the names of features
+    that can be collected.
+
+    Examples:
+
+        >>> model = VGG16()
+        # By default, VGG16.__call__ returns a probability score.
+        >>> prob = model(imgs)
+
+        >>> model.feature_names = 'conv5_3'
+        # This is feature conv5_3.
+        >>> feat5_3 = model(imgs)
+
+        >>> model.feature_names = ['conv5_3', 'fc6']
+        >>> # These are features conv5_3 and fc6.
+        >>> feat5_3, feat6 = model(imgs)
+
+    .. seealso::
+        :class:`chainercv.links.model.SequentialFeatureExtractor`
+
+    When :obj:`pretrained_model` is the path of a pre-trained chainer model
+    serialized as a :obj:`.npz` file in the constructor, this chain model
+    automatically initializes all the parameters with it.
+    When a string in the prespecified set is provided, a pretrained model is
+    loaded from weights distributed on the Internet.
+    The list of pretrained models supported are as follows:
+
+    * :obj:`imagenet`: Loads weights trained with ImageNet and distributed \
+        at `Model Zoo \
+        <https://github.com/BVLC/caffe/wiki/Model-Zoo>`_.
+
+    Args:
+        pretrained_model (str): The destination of the pre-trained
+            chainer model serialized as a :obj:`.npz` file.
+            If this is one of the strings described
+            above, it automatically loads weights stored under a directory
+            :obj:`$CHAINER_DATASET_ROOT/pfnet/chainercv/models/`,
+            where :obj:`$CHAINER_DATASET_ROOT` is set as
+            :obj:`$HOME/.chainer/dataset` unless you specify another value
+            by modifying the environment variable.
+        n_class (int): The dimension of the output of fc8.
+        mean (numpy.ndarray): A mean image. If :obj:`None` and
+            a supported pretrained model is used,
+            the mean image used to train the pretrained model will be used.
```

Comment:
If `pretrained_model == 'imagenet'`, `mean` is not a mean image but a mean value.